Add gsg quick setup script.

lirui34 opened the pull request:
Working out a script to quick setup SOS or UOS.

Co-Authored-By: lirui34 <48583653+lirui34@users.noreply.github.com>

NanlinXie commented:
Can one of the admins verify this patch?

gvancuts commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,207 @@
+
+function print_help()
+{
+echo "Usage:"
```
We should add some more text that explains what this script does. Also making sure that people understand the risks of running this since it will modify various system parameters with full root access.

lirui34 replied:
You are right, I added more text and example usages in it.

gvancuts commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,207 @@
+{
+echo "Usage:"
+echo "Launch this script as: ./flash.sh -t 28100"
+echo "-t to specify clear linux version for upgrading"
```
It would improve readability to add a `tab` at the beginning of each parameter line, e.g. 
```suggestion
echo -e "\t-t to specify clear linux version for upgrading"
```

gvancuts commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,207 @@
+    echo "Add swupd mirror..."
+    swupd mirror -s $mirror
+fi
+unset https_proxy
```
We should only do this if a proxy is provided. Usually, one expects proxy to be transparently. What I mean by that is that if the system knows the proxy information (which should represent 90+% of the cases), then we shouldn't use this setting. This should only be needed where the proxy setting were not automatically determined.

gvancuts replied:
Also, why do you unset this twice (it's also done below in line 108)

lirui34 replied:
Thanks for the comments, it should be removed.

gvancuts commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,207 @@
+    efibootmgr -c -l "\EFI\acrn\acrn.efi" -d /dev/sda -p 1 -L "ACRN"
+
+    echo "Create loader.conf"
+    rm /mnt/loader/loader.conf && touch /mnt/loader/loader.conf
```
How about we at least keep a backup copy of this?
```suggestion
    mv /mnt/loader/loader.conf /mnt/loader/loader.conf.bck && touch /mnt/loader/loader.conf
```

gvancuts commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,207 @@
+sed -i "s/\(virtio-blk.*\)\/home\/clear\/uos\/uos.img/\1$(echo $uos_img_path | sed "s/\//\\\\\//g")/" ~/launch_uos.sh
+echo "Upgrade uos done..."
+echo "Now you can run below command to start UOS..."
+echo "# cd ~/ && ./launch_uos.sh -V 1"
```
We do not have this `-V` parameter in the launch scripts for EFI platforms

lirui34 replied:
Yes, it's reword to this 'cd ~/ && ./launch_uos.sh'

gvancuts commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,207 @@
+    if [[ $? -eq 0 ]]; then
+        losetup -f -P --show clear-$target_version-kvm.img
+    fi
+    mount /dev/loop0p3 /mnt
```
Remove the hardcoded `loop0` here. You need to read it from the previous command, that's what the `--show` flag from `losetup` will do for you, print out the loop device that has been used.

lirui34 replied:
Thanks for the comments, it's updated in latest code, could you please help to review it ?

gvancuts commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,207 @@
+fi
+
+# Do the setups if previous process succeed.
+if [[ $? -eq 0 ]]; then
```
Do we need to reboot after the `swupd verify fix` operation or is it not required.

lirui34 replied:
Yes, it will do reboot after sos setup is complete.

gvancuts commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,207 @@
+    else
+        echo "fail to losetup uos img, please check uos img status" && exit 1
+    fi
+    mount /dev/loop0p3 /mnt
```
Remove the hardcoded loop0 here. You need to read it from the previous command, that's what the --show flag from losetup will do for you, print out the loop device that has been used.

gvancuts replied:
You should also verify that nothing is already mounted under `/mnt`

lirui34 replied:
Yes, it should be handled in latest code.

gvancuts commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,207 @@
+done
+
+# -t opt is must
+if [[ -z $1 ]]; then
```
We should add a check to ensure we have root privileges:
```suggestion
if [[ -z $1 || $EUID -ne 0 ]]; then
```

gvancuts commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,207 @@
+
+function upgrade_uos()
+{
+# Add proxy if you need to download uos img automatically.
```
You never download the UOS image automatically in the script, or did I miss it?

If true, then no need to modify any proxy setting

lirui34 replied:
I have update in this part. Now it is also support to download uos image from official site with proxy settings.

dbkinder commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,207 @@
+echo "-t to specify clear linux version for upgrading"
+echo "-p to specify a proxy server"
+echo "-m to use swupd mirror url"
+echo "-u to upgrade uos, if you just upgrade sos using this script, you'd better to manually reboot sos then continue to use this function."
```
I think I'll wait to do an English edit until after @gvancuts technical issues are addressed. For example this would be better as:
```
echo "-u to upgrade UOS.  After upgrading you must manually reboot the system."
```

lirui34 replied:
Rewrite upgrade uos function. Now it can be upgraded directly without checking sos clear linux version.  Also users don't need to reboot device for taking effects after it.

dbkinder commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,207 @@
+
+function print_help()
+{
+echo "Usage:"
```
Do you have a convention of indenting the body of a function?

lirui34 replied:
I just used tab in it...may I have any suggestion for improvement :smile:

dbkinder replied:
Should this line ``echo "Usage:"`` (and all the script lines in the body of this function) have a tab before them?  I don't see one.

lirui34 replied:
I got it. it will fix in next release.

dbkinder commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,207 @@
+if [[ $upgrade_uos == 1 ]]; then
+    echo "Start upgrading uos to: $target_version ..."
+    upgrade_uos
+    exit
```
Should you write out that the user must reboot the system now, or add a reboot to the script?

lirui34 replied:
Yes, I added text and some of usages in it, could you please help to check in latest code ?

lirui34 commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,207 @@
+swupd autoupdate --disable
+
+# Compare with current clear linux and skip upgrade sos if get the same version.
+if [[ $(echo -e `swupd info` | cut -d' ' -f3) -eq $target_version ]]; then
```
Thanks for your comments, I will update it in next release

lirui34 commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,207 @@
+    echo "Add service-os kernel-iot-lts2018 bundle"
+    swupd bundle-add service-os kernel-iot-lts2018
+
+    echo "mount /dev/sda1 to /mnt"
```
Yes, I have changed code to determine the right ESP, also it will handle multiple ESPs if they are detected via 'fdisk -l' command.

lirui34 replied:
I am now considering to let user determine ESP path(SATA EFI path as default) before running this script just like this:
`# You should set EFI path before upgrading SOS.`
 `efi_partition=/dev/sda1`
` # Check EFI path exists.`
` [[ ! -b /dev/sda1 ]] && echo "You didn't choose the right EFI path." && exit 1`
For the reason, it's hard to auto determine the right ESP on a multiple OS(hard disk) device and also we need this script in automation test framework for some auto flashing cases...but I did not find a good resolution except that to handle such case.

lirui34 commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,207 @@
+    fi
+
+    echo "Add new ACRN efi boot event"
+    efibootmgr -c -l "\EFI\acrn\acrn.efi" -d /dev/sda -p 1 -L "ACRN"
```
Yes, I have used 'fdisk -l' to determine the right ESP, could you please help to review the latest code ?

lirui34 commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,207 @@
+    echo "Check linux bootloader event"
+    number=$(expr `efibootmgr | grep 'Linux bootloader' | wc -l` - 1)
+    if [[ $number -ge 1 ]]; then
+        echo "Clean all Linux bootloader event"
```
I was thinking that people would like to see only one "Linux bootloader" remained in boot list. :smile:

lirui34 commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,207 @@
+    swupd bundle-add service-os kernel-iot-lts2018
+
+    echo "mount /dev/sda1 to /mnt"
+    mount /dev/sda1 /mnt
```
Yes, it will be checked in the latest code.

lirui34 commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,207 @@
+    cur_kernel=`cat /mnt/loader/entries/acrn.conf | sed -n 2p | cut -d'/' -f4`
+
+    echo "Replace root uuid from acrn.conf"
+    sed -i "s/<UUID of rootfs partition>/`blkid -s PARTUUID -o value /dev/sda3`/g" /mnt/loader/entries/acrn.conf
```
Yes, I have updated in latest code, could you please help to review it ?

dbkinder commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,277 @@
+# Compare with current clear linux and skip upgrade sos if get the same version.
+source /etc/os-release
+if [[ $VERSION_ID -eq $target_version ]]; then
+    echo "Specify with the same clear linux version, continue to setup sos..."
```
```suggestion
    echo "Clear Linux version $target_version is already installed. Continuing to setup SOS..."
```

dbkinder commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,277 @@
+if [[ $VERSION_ID -eq $target_version ]]; then
+    echo "Specify with the same clear linux version, continue to setup sos..."
+else
+    echo "Upgrade clear linux version to : $target_version ..."
```
```suggestion
    echo "Upgrading Clear Linux version from $VERSION_ID to $target_version ..."
```

dbkinder commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,277 @@
+        exit 1
+    fi
+    cp -r $acrn_conf_path /mnt/loader/entries/
+    if [[ $? -ne 0 ]]; then echo "You didn't choose an right EFI partition." && exit 1; fi
```
The user can't choose the EFI partition, this is figured out by the script. What's the real error here?

lirui34 replied:
Update message in next release, it's checking whether line: `cp -r $acrn_conf_path /mnt/loader/entries/` is running successfully.

dbkinder commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,277 @@
+
+    echo "Create loader.conf"
+    mv /mnt/loader/loader.conf /mnt/loader/loader.conf.bck && touch /mnt/loader/loader.conf
+    echo "Add default boot wait time"
```
```suggestion
    echo "Add default (5 seconds) boot wait time"
```

dbkinder commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,277 @@
+    echo "Getting current Service OS kernel version: $cur_kernel"
+
+    echo "Replacing root partition uuid in acrn.conf"
+    sed -i "s/<UUID of rootfs partition>/`blkid -s PARTUUID -o value $root_partition`/g" /mnt/loader/entries/acrn.conf
```
You should add a test to check this command didn't fail.

lirui34 replied:
OK, I will add some test in next release.

dbkinder commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,277 @@
+
+umount /mnt
+sync
+reboot
```
If the upgrade fails, you're still rebooting the system?  The user may not get a chance to see the failure error message.

lirui34 replied:
Add several tests for the previous processing of upgrade SOS, if any one of them fail, the script will stopped so that it will not reboot device.
Also I have added disable_reboot parameter to let user skip rebooting after upgrade.

dbkinder commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,277 @@
+        uos_img=`echo $uos_img_xz | sed 's/.xz$//g'`
+    fi
+    echo "get uos img file: $uos_img"
+    if [[ $? -eq 0 ]]; then
```
This test of ``$?`` will return the result of the ``echo`` command execution, right?  So it will always be 0.

lirui34 replied:
Yes, it's removed in next release.

dbkinder commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,277 @@
+    curl $uos_image_link -o clear-$target_version-kvm.img.xz
+    uos_img=clear-$target_version-kvm.img
+    if [[ -f $uos_img ]] && [[ -f $uos_img.xz ]]; then echo "Moving $uos_img to $uos_img.old."; mv $uos_img $uos_img.old; fi
+    if [[ $? -eq 0 ]]; then
```
Which command execution status do you think this it checking?   It's not the ``curl``

lirui34 replied:
Remove this check event in next release.

dbkinder commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,277 @@
+        echo "Unxz UOS image: clear-$target_version-kvm.img.xz"
+        unxz clear-$target_version-kvm.img.xz
+    fi
+    if [[ $? -eq 0 ]]; then
```
I'm not sure which command execution status this checks either...

lirui34 replied:
Also remove it in next release.

dbkinder commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,277 @@
+cp -r /usr/share/acrn/samples/nuc/launch_uos.sh ~/
+sed -i "s/\(virtio-blk.*\)\/home\/clear\/uos\/uos.img/\1$(echo $uos_img | sed "s/\//\\\\\//g")/" ~/launch_uos.sh
+echo "Upgrade UOS done..."
+echo "Now you can run below command to start UOS..."
```
```suggestion
echo "Now you can run this command to start UOS..."
```

gvancuts commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,277 @@
+    export https_proxy=$proxy
+fi
+
+if [[ ! -z `findmnt /mnt -n` ]]; then umount /mnt; fi
```
I think it's a bit rude of the script to silently unmount `/mnt` if it is already mounted. Perhaps the user is using this for some other purpose. It would be better to exit gracefully and ask him to free it up himself.

lirui34 replied:
Yes, it's fixed.

gvancuts commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,277 @@
+    cp -r /usr/lib/modules/"`readlink /usr/lib/kernel/default-iot-lts2018 | awk -F '2018.' '{print $2}'`.iot-lts2018" /mnt/lib/modules
+fi
+umount /mnt
+losetup -D
```
The user may be using other loop devices for other things, we shouldn't detach *all* of them unconditionally.

lirui34 replied:
Removed this operation

lirui34 commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,277 @@
+# since it will modify various system parameters with full root access.
+#
+# Usages:
+#   Upgrade sos to 28100
```
It will update in next release.

lirui34 commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,277 @@
+#   Upgrade sos to 28100 with specified proxy server
+#     sudo <script> -t 28100 -p <your proxy server>:<port>
+#   Upgrade uos to 28100 with specified proxy server
+#     sudo <script> -t 28100 -p <your proxy server>:<port> -u
```
Agree, I'll move it to the first option

lirui34 commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,277 @@
+echo -e "\t-t to specify clear linux version for upgrading"
+echo -e "\t-p to specify a proxy server"
+echo -e "\t-m to use swupd mirror url"
+echo -e "\t-u to upgrade UOS."
```
I changed usage to: -s to upgrade SOS and -u to upgrade UOS.  and move original -s to -k for skip download, add -d to disable auto reboot after upgrade SOS.

lirui34 commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,277 @@
+        esac
+done
+
+# -t opt is must
```
Changed options check methods.

lirui34 commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,277 @@
+                        ;;
+        esac
+done
+
```
Yes, this line will check if script is running with `sudo`:
`[[ $EUID -ne 0 ]] && echo "You have to run script as root." && exit 1`

lirui34 commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,277 @@
+
+# Exit script if you specified an newer clear linux version.
+if [[ $(echo -e `swupd info` | cut -d' ' -f3) -gt $target_version ]]; then
+    echo -e 'No need to upgrade Service OS.' && exit
```
Yes, update message in next release.

lirui34 commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,277 @@
+if [[ -z $efi_partition ]]; then echo "You didn't choose a partition, please try it again."; exit 1; fi
+efi_mount_point=`findmnt $efi_partition -n | cut -d' ' -f1`
+root_partition=`echo $efi_partition | sed 's/[0-9]$/3/g'`
+partition=`echo $efi_partition | sed 's/[0-9]$//g'`
```
Fixed

dbkinder commented:
@NanlinXie who should do the final approval of this PR and merge?

gvancuts commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,246 @@
+    # You should set EFI path before upgrading SOS.
+    efi_partition=/dev/sda1
+    # Check EFI path exists.
+    [[ ! -b $efi_partition ]] && echo "Please choose the right EFI path firstly." && exit 1
```
How can the user set the ESP to the right value? Is editing the script the only option?

We should add a note explaining this

lirui34 replied:
Yes, so far it's the only option to edit script to set ESP.
I have also added a note at the head of the script to explain user should set the right ESP before running script.

gvancuts replied:
We should really make this a command-line parameter and the explanation should be included in the "usage" section. People do not typically read scripts before running them. It's great to have details and explanations in them but for something that really has to be provided by the user, a parameter should be used (as opposed to editing the script).

lirui34 replied:
Thank you very much for suggestion, I have added a new parameter "-e" for user to specify the right ESP and also added "note" into print_help function.

gvancuts commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,246 @@
+    if [[ -n $efi_mount_point ]]; then
+        echo "EFI partition is mounted: $efi_mount_point..."
+        echo "You should unmount it before upgrading SOS."
+        exit 1
```
We do not need to error out, we could simply use that mount point directly, couldn't we?

lirui34 replied:
Yes, we can use it directly.

gvancuts commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,246 @@
+
+    # Do the setups if previous process succeed.
+    if [[ $? -eq 0 ]]; then
+        echo "Add service-os kernel-iot-lts2018 bundle"
```
```suggestion
        echo "Adding the service-os and kernel-iot-lts2018 bundles..."
```

NanlinXie commented:
@gvancuts , Do you have more comment for the script?

gvancuts commented on doc/getting-started/gsg_quick_setup.sh:
```diff
@@ -0,0 +1,248 @@
+    echo "Launch this script as: sudo $0 -s 28100"
+    echo -e "\t-s to upgrade SOS"
+    echo -e "\t-u to upgrade UOS"
+    echo -e "\t-p to specify a proxy server"
```
Nitpick:
```suggestion
    echo -e "\t-p to specify a proxy server (HTTPS)"
```

lirui34 commented:
> Very close!
Hi @gvancuts , I have added a new parameter "-e" for user to specify their ESP and also added "note" into print_help function.
Could you please have a look at those changes ?
